storage: introduce metrics for intent resolutions

tbg opened the pull request:
In particular, this tracks poisoning aborts, which create abort span
entries (see #25233).

Some refactoring to make it easier to add such counters in the future
was carried out.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25355)
<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/client_metrics_test.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-LC-xeWuDT02Nve9Y-tK:-LC-xeWuDT02Nve9Y-tL:b-xlousg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60dc764faa55f7a85aaae5edcb2a4e91390d0ef/pkg/storage/client_metrics_test.go#L167)):*
> ```Go
> 		act := fmt.Sprintf("%+v", result.Metrics{})
> 		exp := "{LeaseRequestSuccess:0 LeaseRequestError:0 LeaseTransferSuccess:0 LeaseTransferError:0 ResolveCommit:0 ResolveAbort:0 ResolvePoison:0}"
> 		if act != exp {
> ```

nit: probably cleaner to use [`reflect.Value.NumFields`](https://golang.org/pkg/reflect/#Value.NumField).

---

*[pkg/storage/metrics.go, line 971 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-LC-y6P64Wmgpie6K2qV:-LC-y6P64Wmgpie6K2qW:b-avxjg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60dc764faa55f7a85aaae5edcb2a4e91390d0ef/pkg/storage/metrics.go#L971)):*
> ```Go
> 	metric.ResolvePoison = 0
> 
> 	if metric != (result.Metrics{}) {
> ```

I like this pattern. Glad you're making it more prolific.

---

*[pkg/storage/batcheval/result/result_test.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-LC-ziZ4DqxMsCaKuwWX:-LC-ziZ4DqxMsCaKuwWY:bzcs61c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60dc764faa55f7a85aaae5edcb2a4e91390d0ef/pkg/storage/batcheval/result/result_test.go#L71)):*
> ```Go
> 	r2.Local.Metrics = new(Metrics)
> 
> 	*r1.Local.Metrics = LeaseRequestSuccess
> ```

Does this work? I thought this was no longer an enum. I must be missing something.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-:-LC-zwlb6c9NA7FL93El:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR!

bors r=nvanbenschoten

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-:-LJxqbEg5pvrU3UUSSqJ:b-l0r5kv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_metrics_test.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-LC-xeWuDT02Nve9Y-tK:-LJxmBgG3JVY6m6tKCj3:b-us2nvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60dc764faa55f7a85aaae5edcb2a4e91390d0ef/pkg/storage/client_metrics_test.go#L167)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: probably cleaner to use [`reflect.Value.NumFields`](https://golang.org/pkg/reflect/#Value.NumField).
</blockquote></details>

That won't catch removal of a field and addition of one at the same time. We change the names so rarely that I think this is fine.
___
*[pkg/storage/metrics.go, line 971 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-LC-y6P64Wmgpie6K2qV:-LJxmI9g685OAOJtHR7J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60dc764faa55f7a85aaae5edcb2a4e91390d0ef/pkg/storage/metrics.go#L971)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I like this pattern. Glad you're making it more prolific.
</blockquote></details>

Done.
___
*[pkg/storage/batcheval/result/result_test.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25355#-LC-ziZ4DqxMsCaKuwWX:-LJxq_jJ1l3UcJgS8NYm:b9mrckg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60dc764faa55f7a85aaae5edcb2a4e91390d0ef/pkg/storage/batcheval/result/result_test.go#L71)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this work? I thought this was no longer an enum. I must be missing something.
</blockquote></details>

No, this was broken. Fixed.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: introduce metrics for intent resolutions

In particular, this tracks poisoning aborts, which create abort span
entries (see #25233).

Some refactoring to make it easier to add such counters in the future
was carried out.

Release note: None

tschottdorf commented:
Flake filed as https://github.com/cockroachdb/cockroach/issues/28694.

bors r=nvanbenschoten
